traverse Array with Maybe more quickly

andrewthad opened the pull request:
WIP. I still need to handle `SmallArray`. I also need to make this work with `Either`, but the backwards-compatibility story there is more annoying because of the historical absence of `ExceptT`. I also want to do this for `State`. Probably not `Writer` though since it has abysmal performance anyway.

treeowl commented:
I like this. For `Either`, you can either copy the basics of `ExceptT` from `transformers` or (perhaps better) just manually work it into the code for `traverseArrayP`.

Side note: it's most unfortunate for `RULES` that we don't have, e.g., `Applicative (ExceptT e m) :- Applicative m`, and that we have no way to match on the use of a particular instance dictionary either. So all these rules have to be pretty much monomorphic.

treeowl commented:
I don't understand why you think the manual copies (with rewrite rules to match) wouldn't help.

andrewthad commented:
I think GHC will produce the same core.

treeowl commented:
Quite possibly, but you won't be depending on having ExceptT in
transformers.

On Apr 22, 2018 10:58 AM, "Andrew Martin" <notifications@github.com> wrote:

> I think GHC will produce the same core.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/primitive/pull/142#issuecomment-383387491>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABzi_VIxxCgLXNeO_AbANLNZKCV4NYIzks5trJqYgaJpZM4Tdr0z>
> .
>

andrewthad commented:
Ah, it was consistent behavior with all versions of containers that was your concern. I've changed it as you suggested. Also, I added a benchmark just to see if there was a performance difference between the two implementations. They have identical performance (Wooh GHC!).

andrewthad commented:
> All these rewrite rules (including the ones I wrote) feel somewhat unsatisfactory, because they completely fall down for transformers. I'm wondering if there might be a way to catch known transformations of known base monads, using rewrite rules with fall-backs. 

I can't find a way to make this work. Consider what a rewrite rule for `MaybeT` might look like:

   "traverse/MaybeT" forall (f :: a -> MaybeT m b). traverseArray f = ...

Here's the problem: in the rewrite rule, we have no way to tell GHC that `m` needs to have a `PrimMonad` constraint. GHC appears to have a syntax that tricks the user into thinking they can do this, but it doesn't actually work.

andrewthad commented:
In fact, in the rewrite rule, GHC doesn't even know that `m` has an `Applicative` instance.

treeowl commented:
You certainly can't do it that way. I'm not sure if you can do it at all or
not, but if so it will require rewrite rules "all the way down" to a base
type a rule matches on.

On Mon, Apr 23, 2018, 3:36 PM Andrew Martin <notifications@github.com>
wrote:

> All these rewrite rules (including the ones I wrote) feel somewhat
> unsatisfactory, because they completely fall down for transformers. I'm
> wondering if there might be a way to catch known transformations of known
> base monads, using rewrite rules with fall-backs.
>
> I can't find a way to make this work. Consider what a rewrite rule for
> MaybeT might look like:
>
> "traverse/MaybeT" forall (f :: a -> MaybeT m b). traverseArray f = ...
>
> Here's the problem: in the rewrite rule, we have no way to tell GHC that m
> needs to have a PrimMonad constraint. GHC appears to have a syntax that
> tricks the user into thinking they can do this, but it doesn't actually
> work.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/primitive/pull/142#issuecomment-383696469>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABzi_dZCqTaNqZmbf1VIp02zVy3_B4Aaks5tri04gaJpZM4Tdr0z>
> .
>

treeowl commented:
The basic idea would be to step through the layers of monad transformers
you recognize, recording how each one transforms a monad. Then at the
bottom, find a base monad you recognize and build up the ultimate bind and
return. I'm not sure this can be done, but I strongly suspect it can.

On Mon, Apr 23, 2018, 3:43 PM Andrew Martin <notifications@github.com>
wrote:

> In fact, in the rewrite rule, GHC doesn't even know that m has an
> Applicative instance.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/primitive/pull/142#issuecomment-383698478>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABzi_bT_4FDHXBWsf7sCOMq0KZFsl94lks5tri7igaJpZM4Tdr0z>
> .
>

andrewthad commented:
I'm still not able to follow how this is supposed to look. Going back to `MaybeT`:

    "traverse/MaybeT" forall (f :: a -> MaybeT m b). traverseArray f = ...

I cannot see anything that could go on the RHS (other than `traverseArray f`) that is both correct and satisfies the type checker.

andrewthad commented:
Wait, I think I might be beginning to see a way.

treeowl commented:
Wait, I think I might be beginning to see a way.
>

Me too 😉

> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/primitive/pull/142#issuecomment-383705732>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABzi_UxLd-9G6tLrM5RAYiQkfb2cd6_xks5trjSQgaJpZM4Tdr0z>
> .
>

treeowl commented:
The race is on.

treeowl commented:
I think we can get MaybeT, ExceptT, and StateT, but not WriterT, RWST, or
ErrorT (since those have side constraints). I haven't actually tested my
rules, though, so I don't know if they fire or, if they do, whether they
produce decent code.

On Mon, Apr 23, 2018, 4:12 PM David Feuer <david.feuer@gmail.com> wrote:

>
> Wait, I think I might be beginning to see a way.
>>
>
> Me too 😉
>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/haskell/primitive/pull/142#issuecomment-383705732>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABzi_UxLd-9G6tLrM5RAYiQkfb2cd6_xks5trjSQgaJpZM4Tdr0z>
>> .
>>
>

andrewthad commented:
Have you pushed these somewhere?

treeowl commented:
No. Thus far they're pretty much faked up. When I'm home I'll (try to) make
them a bit less fake and push to show you.

On Mon, Apr 23, 2018, 4:53 PM Andrew Martin <notifications@github.com>
wrote:

> Have you pushed these somewhere?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/primitive/pull/142#issuecomment-383719183>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABzi_aDZfEmPXOF-IzFWAsJpLuOxWDjDks5trj9TgaJpZM4Tdr0z>
> .
>

treeowl commented:
Until then, I strongly urge you to give it a good try. You may well come up
with a better idea than I did, and in any case you'll certainly learn some
things.

On Mon, Apr 23, 2018, 5:05 PM David Feuer <david.feuer@gmail.com> wrote:

> No. Thus far they're pretty much faked up. When I'm home I'll (try to)
> make them a bit less fake and push to show you.
>
> On Mon, Apr 23, 2018, 4:53 PM Andrew Martin <notifications@github.com>
> wrote:
>
>> Have you pushed these somewhere?
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/haskell/primitive/pull/142#issuecomment-383719183>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABzi_aDZfEmPXOF-IzFWAsJpLuOxWDjDks5trj9TgaJpZM4Tdr0z>
>> .
>>
>

cartazio commented:
It feels like we’re bordering on feature / ideas for improving rules based
optimization. What is it we wish we could say / can only express via hand
scripted hermit tool optimization’s?

David: have you tried out hermit?

On Mon, Apr 23, 2018 at 9:21 PM David Feuer <notifications@github.com>
wrote:

> Hrmm... I'm able to write code that at least type checks to handle (some)
> transformers on top of ST and IO, but I'm actually not sure how to handle
> transformers on top of Maybe, Either, Identity, etc. Do you think it's
> doable? If not for the annoying limits on instance discovery, do you think
> you could write rules? Maybe those could inspire. Anyway, I'll upload what
> I have within the hour.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/primitive/pull/142#issuecomment-383772437>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQwoT8qU8bQdwXqJnxjl4erPb-ylDkks5trn4jgaJpZM4Tdr0z>
> .
>

treeowl commented:
@cartazio, I've only looked at hermit briefly. It's rather complicated. The basic limitation is basically that the instance resolution derivations drop away after type checking. So for example the simplifier doesn't know how GHC resolved `PrimMonad (WriterT w m)` and therefore doesn't know `Monad m` or `Monoid w`.

cartazio commented:
Is there ways this could be changed in ghc?

Seems like it would be at least a useful discussion as feature request
motivated by what optimization’s you want to be able to say simply.

I’d like it if we can keep primitive simple if we can.  At some point too
much rules optImitation can hinder easy of changing internals.  Though we
don’t have much internals going on :)

On Mon, Apr 23, 2018 at 11:50 PM David Feuer <notifications@github.com>
wrote:

> @cartazio <https://github.com/cartazio>, I've only looked at hermit
> briefly. It's rather complicated. The basic limitation is basically that
> the instance resolution derivations drop away after type checking. So for
> example the simplifier doesn't know how GHC resolved PrimMonad (WriterT w
> m) and therefore doesn't know Monad m or Monoid w.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/primitive/pull/142#issuecomment-383794951>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQwp1B8vcjC9gp9iAxWzcxqVfS4i63ks5trqD8gaJpZM4Tdr0z>
> .
>

treeowl commented:
For stacks based on `ST` or `IO`, my version seems *much* simpler, and involves far fewer enormous piles of code that need to go away. I suggest you just ignore those stacks altogether and try to come up with a solution for stacks based on `Maybe`, `Either`, `Identity`, etc. It should be easy to use both solutions at the same time.

treeowl commented:
Separately, I'm considering a simpler general version that might work somewhat better than the one I put in before. In particular, we can go back to something somewhat list-like, but using a list that holds multiple elements per cons. I need to do some benchmarking.

andrewthad commented:
Undoubtedly, yours is much simpler.  In my most recent commit, I was mostly playing around just to see  what it takes to make things like Maybe and Either accepted as the base monad.  It’s helped me understand the problem better.  I  have another idea for how to extend  your solution to accept  different base monads  that I am going to try soon.

treeowl commented:
@cartazio, the fast way to traverse an array is to

1. Create a mutable array.
2. For each element of the given array,
    a. Perform an action to get an element and
    b. Write that element to the mutable array.
3. Freeze the mutable array and return the resulting immutable array.

We'd like `traverse` to do that whenever it can, but it's not always possible. Even ignoring safety, note that `Compose m n` is not generally a `PrimMonad`, or even a `Monad`, even if `m` and `n` both are.

cartazio commented:
I'm not terribly familiar with Compose, could you walk me through an example?

cartazio commented:
also where does Compose live?

cartazio commented:
hrmm

if this is the correct defintion of compose

```
-- | Right-to-left composition of functors.
-- The composition of applicative functors is always applicative,
-- but the composition of monads is not always a monad.
newtype Compose f g a = Compose { getCompose :: f (g a) }

```
wouldn't  just require PrimMonad just one of these?

cartazio commented:
hrmmm, that makes me think that we can't have an instance for compose, because there should only be one primonad in the stack ... i think?

cartazio commented:
(something something global approximate  linearity of state token)

cartazio commented:
i guess i'm perhaps not understnding why Compose the data type/monad enters into this, but i'll have to look at more details

treeowl commented:
Not even compositions of pure monads are necessarily monads. As my PR
shows, we can do good things for a lot of transformer stacks built on ST
and IO. Andrew has been working on an idea for stacks built on Identity,
Either, and Maybe  (and perhaps even (a,)), but it's not ready yet. Compose
is a generally useful way to construct Applicatives, which people certainly
use with traverse, but I don't know if we'll be able to optimize for it or
not.

On Fri, Apr 27, 2018, 5:45 PM Carter Tazio Schonwald <
notifications@github.com> wrote:

> (something something global approximate linearity of state token)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/primitive/pull/142#issuecomment-385102244>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABzi_RalzTkE83YEyH1yLdW5ZDNst80mks5ts5GAgaJpZM4Tdr0z>
> .
>

treeowl commented:
The composition law for Traversable does seem to suggest we can probably do
something smart about Compose, actually. We know that

traverse
<http://hackage.haskell.org/package/base-4.11.1.0/docs/Data-Traversable.html#v:traverse>
(Compose
. fmap
<http://hackage.haskell.org/package/base-4.11.1.0/docs/Data-Functor.html#v:fmap>
g
. f) = Compose . fmap
<http://hackage.haskell.org/package/base-4.11.1.0/docs/Data-Functor.html#v:fmap>
 (traverse
<http://hackage.haskell.org/package/base-4.11.1.0/docs/Data-Traversable.html#v:traverse>
g)
. traverse
<http://hackage.haskell.org/package/base-4.11.1.0/docs/Data-Traversable.html#v:traverse>
 f

In particular,

traverse (Compose . f) = Compose . fmap sequenceA . traverse f

So we can turn one traversal into two, which sounds bad but is probably
actually good. For *pure* Applicatives, I conjecture that we can also turn
two traversals into one efficient one. Anyway, that's the start of a
thought in that direction. Still need to think it through a bit more.

On Fri, Apr 27, 2018, 6:01 PM David Feuer <david.feuer@gmail.com> wrote:

> Not even compositions of pure monads are necessarily monads. As my PR
> shows, we can do good things for a lot of transformer stacks built on ST
> and IO. Andrew has been working on an idea for stacks built on Identity,
> Either, and Maybe  (and perhaps even (a,)), but it's not ready yet. Compose
> is a generally useful way to construct Applicatives, which people certainly
> use with traverse, but I don't know if we'll be able to optimize for it or
> not.
>
> On Fri, Apr 27, 2018, 5:45 PM Carter Tazio Schonwald <
> notifications@github.com> wrote:
>
>> (something something global approximate linearity of state token)
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/haskell/primitive/pull/142#issuecomment-385102244>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABzi_RalzTkE83YEyH1yLdW5ZDNst80mks5ts5GAgaJpZM4Tdr0z>
>> .
>>
>
